trie/node.go

```diff
@@ -185,6 +187,127 @@ func (l *leaf) Encode() ([]byte, error) {
 	return encoding, nil
 }
 
+func Decode(r io.Reader) (node, error) {
+	header, err := readByte(r)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeType := header >> 6
+	if nodeType == 1 {
+		l := new(leaf)
+		err := l.Decode(r, header)
+		return l, err
+	} else if nodeType == 2 || nodeType == 3 {
+		b := new(branch)
+		err := b.Decode(r, header)
+		return b, err
+	}
+
+	return nil, errors.New("cannot decode invalid encoding into node")
+}
+
+// Decode decodes a byte array with the following format into a branch node
+// NodeHeader | Extra partial key length | Partial Key | Value
+// where NodeHeader is a byte:
+// most significant two bits of first byte: 10 if branch w/o value, 11 if branch w/ value
+// least significant six bits of first byte: if len(key) > 62, 0x3f, otherwise len(key)
+// where Extra partial key length is included if len(key) > 63: consists of the remaining key length
+// note that partial key length is the length of the pk in nibbles
+// Partial Key is the branch's key
+// Value is:
+// Children Bitmap | SCALE Branch Node Value | Hash(Enc(Child[i_1])) | Hash(Enc(Child[i_2])) | ... | Hash(Enc(Child[i_n]))
+// Note that since the encoded branch stores the hash of the children nodes, we aren't able to reconstruct the child
+// nodes from the encoding. This function instead stubs where the children are known to be with an empty leaf.
+func (b *branch) Decode(r io.Reader, header byte) (err error) {
+	if header == 0 {
```

Comment:
in which case would we have header == 0?

Reply:
if the header hasn't been read yet, if we pass in a value for header it means we already read it from the reader, so we don't want to read another byte for it